zipline/pipeline/loaders/blaze.py

```diff
@@ -0,0 +1,904 @@
+            )
+    return dict(adjustments)  # no subclasses of dict
+
+
+class BlazeLoader(dict):
+    def __init__(self, colmap=None):
+        self.update(colmap or {})
+
+    @classmethod
+    @memoize(cache=WeakKeyDictionary())
+    def global_instance(cls):
+        return cls()
+
+    def __hash__(self):
+        return id(self)
+
+    def __call__(self, column):
+        if column.dataset in self:
+            return self
+        raise KeyError(column)
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        return map(
+            op.getitem(
+                dict(concat(
+                    self._load_dataset(cs, dates, assets, mask)
+                    for _, cs in iteritems(groupby(getdataset, columns))
+                )),
+            ),
+            columns,
+        )
+
+    def _load_dataset(self, columns, dates, assets, mask):
+        try:
+            dataset, = set(map(getdataset, columns))
+        except ValueError:
+            raise AssertionError('all columns must come from the same dataset')
+
+        expr, deltas, resources = self[dataset]
+        have_sids = SID_FIELD_NAME in expr.fields
+        assets = list(map(int, assets))  # coerce from numpy.int64
+        fields = list(map(dataset_name, columns))
+        query_fields = fields + [AD_FIELD_NAME, TS_FIELD_NAME] + (
+            [SID_FIELD_NAME] if have_sids else []
+        )
+
+        def where(e):
+            """Create the query to run against the resources.
+
+            Parameters
+            ----------
+            e : Expr
+                The baseline or deltas expression.
+
+            Returns
+            -------
+            q : Expr
+                The query to run.
+            """
+            ts = e[TS_FIELD_NAME]
+            # Hack to get the lower bound to query:
+            # This must be strictly executed because the data for `ts` will
+            # be removed from scope too early otherwise.
+            lower = odo(ts[ts <= dates[0]].max(), pd.Timestamp)
+            selection = ts <= dates[-1]
+            if have_sids:
+                selection &= e[SID_FIELD_NAME].isin(assets)
+            if lower is not pd.NaT:
+                selection &= ts >= lower
+
+            return e[selection][query_fields]
+
+        extra_kwargs = {'d': resources} if resources else {}
+        materialized_expr = odo(where(expr), pd.DataFrame, **extra_kwargs)
+        materialized_deltas = (
+            odo(where(deltas), pd.DataFrame, **extra_kwargs)
+            if deltas is not None else
+            pd.DataFrame(columns=query_fields)
+        )
+
+        # Inline the deltas that changed our most recently known value.
+        # Also, we reindex by the dates to create a dense representation of
+        # the data.
+        sparse_output = inline_novel_deltas(
+            materialized_expr,
+            materialized_deltas,
+            dates,
+        ).drop(AD_FIELD_NAME, axis=1).set_index(TS_FIELD_NAME)
```

Comment:
We also drop `AD_FIELD_NAME` in `inline_novel_deltas`.  Can we do that just in one place?